src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -306,29 +273,51 @@ public static ImmutableArray<DocumentSymbolUIItem> GetDocumentSymbolUIItems(Immu
         }
 
         /// <summary>
-        /// Updates the IsExpanded property for the Document Symbol UI representation based on the given Expansion Option. The parameter
+        /// Updates the IsExpanded property for the Document Symbol ViewModel based on the given Expansion Option. The parameter
         /// <param name="currentDocumentSymbolItems"/> is used to reference the current node expansion in the view.
         /// </summary>
-        public static void SetIsExpanded(
-            ImmutableArray<DocumentSymbolUIItem> documentSymbolItems,
-            IEnumerable<DocumentSymbolUIItem> currentDocumentSymbolItems,
+        public static void SetIsExpandedOnNewItems(
+            ImmutableArray<DocumentSymbolDataViewModel> newDocumentSymbolItems,
+            IEnumerable<DocumentSymbolDataViewModel> currentDocumentSymbolItems)
+        {
+            using var _ = PooledHashSet<DocumentSymbolDataViewModel>.GetInstance(out var hashset);
+            hashset.AddRange(newDocumentSymbolItems);
+
+            foreach (var item in currentDocumentSymbolItems)
+            {
+                if (!hashset.TryGetValue(item, out var newItem))
+                {
+                    continue;
+                }
+
+                newItem.IsExpanded = item.IsExpanded;
```

Comment:
would like docs/asserts on the method about threading.

Reply:
I don't understand. How do we assert that we have no thread afinity? Do we want to doc how booleans work with threading?

Reply:
so it can either be a comment at hte start of hte method saying this is expected to be called on any thread.  Or it can be explicit calls to things like `ThrowIfNotOnBackgroundThread`, or `ThisCanBeCalledOnAnyThread`.  it basically just serves as documentation to future readers, which is esp. helpful for multithreaded code with shared concurrency. :)

Reply:
`ThisCanBeCalledOnAnyThread` only applies to types inheriting from [ForegroundThreadAffinitizedObject](https://github.com/dotnet/roslyn/blob/6a347245519d3083297a12b34b1498f5d22bd942/src/EditorFeatures/Core/Shared/Utilities/ForegroundThreadAffinitizedObject.cs#L67). It sounds like you are saying that all types in roslyn should call this at all callsites for clarity?

Reply:
I can add an empty extension method to `IThreadingContext` called `ThisCanBeCalledOnAnyThread`

Reply:
> It sounds like you are saying that all types in roslyn should call this at all callsites for clarity?

for types that are intended to be used in a concurrent and mutable fashion, with orchestration around threading, yes.  :)

Types that aren't mutable don't need it.  Types that are about managing the interaction between a model and a view though very much do, and we do try to follow that in lots of our impls (and it helps catch a bunch of problems :)).

Reply:
I don't see much of a different between calling a method that does nothing but tells the viewer its safe for this to run on any thread and just having a comment stating as such